### DISCUSSION TITLE
Add "Favorite Steps" Functionality

### DISCUSSION THREAD

**Discussion by maciekpazola** [reqnroll/Reqnroll#331] [Ideas]
It would be extremely helpful to have a feature in Reqnroll that allows users to mark certain step definitions as "favorites." This functionality would streamline the process of writing feature files, making it quicker and easier to reuse commonly used steps.

Proposed Functionality:

1. Add the ability to mark specific steps as "favorites" within the Reqnroll IDE plugin.
2. Provide a dedicated section or dropdown in the editor where these favorite steps can be viewed and quickly inserted into feature files.
3. (Optional) Allow organizing favorites into categories for better management.

Use Case:
When working on large projects, there are often a handful of steps that are reused across multiple feature files. Having the ability to quickly access these steps without searching through existing files or the step definitions would save time and improve efficiency.

Benefits:

- Streamlines the process of writing feature files.
- Reduces the likelihood of inconsistencies or errors when retyping steps.
- Improves user productivity, especially on projects with repetitive or complex step definitions.

Thank you for considering this enhancement

**Comment by gasparnagy** [comment]
@maciekpazola This is interesting indeed, but I don't think we have the technical possibilities to do that.

Currently we use Visual Studio's built-in completion infrastructure and regarding my memories that does not provide any options to do prioritizations, sections or special highlights. Replacing that with a robust-enough custom solution is beyond our options, I guess.

Nevertheless you can have a look at the infrastructure we use (see https://github.com/reqnroll/Reqnroll.VisualStudio/tree/main/Reqnroll.VisualStudio/Editor/Completions/Infrastructure), play with it a little bit and if you find any good option to add such highlighting, we can think about how we could achieve that.

**Reply by maciekpazola** [reply]
Hi @gasparnagy,
I reviewed the task and created a small proof of concept. The final solution will depend on the specific requirements we aim to achieve.

Here are the changes I implemented:

- Sorting: Steps are now sorted by their usage frequency within the project.
- DisplayText: The menu display now includes the usage count for each step.
- Icons: Added an icon for the top 5 most frequently used steps.
- Descriptions: For the top 5 steps, the description is set to "⭐ Frequently used step", while for the rest it is "Common step".

Let me know your thoughts or if further adjustments are needed!
![image](https://github.com/user-attachments/assets/7e7600e3-3b52-40de-b6de-44a43d2842ff)

**Reply by gasparnagy** [reply]
This looks cool, but how do you count the number of usages? Or is it only the number of usages in the file?

**Reply by maciekpazola** [reply]
> This looks cool, but how do you count the number of usages? Or is it only the number of usages in the file?

The **StepDefinitionUsageFinder** class already had a method named **FindUsages**, which returned all usages of steps in feature files within the project context, albeit with some minor limitations. To address these limitations, I added a new method to the class that handles edge cases and returns the usages as a Dictionary<string, int>, where the key is the step text and the value is the number of usages in the project.

Subsequently, I modified the **GetStepCompletions** method in the **DeveroomCompletionSource** class to retrieve the usage data for all steps. I then sorted the results and added an icon to the top 5 steps based on their usage count. Additionally, I made adjustments to ensure that the number of usages is not displayed within the step itself but is instead moved to the step's description (displayed as a tooltip when a step is selected in the context menu).

This week I will provide a pull request, but first I've to fix couple tests.

**Reply by maciekpazola** [reply]
@gasparnagy Okay, so I finished all the tasks and did some cleanup, but there's an issue with the StepCompletion tests. To get the completion with the correct number of usages, I had to change the approach for retrieving data. Previously, it was fetched from the sampler (a sort of caching system), but now it retrieves the data directly from the actual files.

To summarize, my implementation is correct, but this change in logic has caused 12 tests to fail. What should we do about this?

I think the simplest solution might be to create a feature file that includes the actual steps within this step. What do you think?
![image](https://github.com/user-attachments/assets/1450b065-88e9-4da5-af9a-181e67d79259)

Pull request:
https://github.com/reqnroll/Reqnroll.VisualStudio/pull/56
